File: LeetCode/0547. Friend Circles/Solution.kt
/**
 * Created by Inno Fang on 2018/3/26.
 */

/**
 * 113 / 113 test cases passed.
 * Status: Accepted
 * Runtime: 448 ms
 */
class Solution {

    class UnionFind(var group: Int) {
        val list = IntArray(group) { it }
        val rank = IntArray(group) { 1 }
        fun find(x: Int): Int {
            if (x != list[x]) list[x] = find(list[x])
            return list[x]
        }

        fun union(x: Int, y: Int) {
            val fx = find(x)
            val fy = find(y)
            if (fx == fy) return
            when {
                rank[fx] < rank[fy] -> list[fx] = fy
                rank[fx] > rank[fy] -> list[fy] = fx
                else -> {
                    list[fx] = fy
                    rank[fx]++
                }
            }
            group--
        }
    }

    fun findCircleNum(M: Array<IntArray>): Int {
        val n = M.size
        val uf = UnionFind(n)
        for (i in (0 until n)) {
            for (j in (i + 1 until n)) {
                if (M[i][j] == 1) uf.union(i, j)
            }
        }
        return uf.group
    }
}

fun main(args: Array<String>) {
//    Solution().findCircleNum(arrayOf(intArrayOf(1, 1, 0), intArrayOf(1, 1, 0), intArrayOf(0, 0, 1))).let(::println)
    Solution().findCircleNum(arrayOf(intArrayOf(1, 1, 1), intArrayOf(1, 1, 1), intArrayOf(1, 1, 1))).let(::println)
}

File: LeetCode/0547. Friend Circles/solution.py
"""
113 / 113 test cases passed.
Runtime: 48 ms
Memory Usage: 15.7 MB
"""
class Solution:
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        ans, n = 0, len(isConnected)
        vis = [False] * n
        def dfs(curr):
            for i in range(n):
                if isConnected[curr][i] and not vis[i]:
                    vis[i] = True
                    dfs(i)
        for i in range(n):
            if not vis[i]:
                dfs(i)
                ans += 1
        return ans

"""
113 / 113 test cases passed.
Runtime: 48 ms
Memory Usage: 15.4 MB
"""
class Solution2:
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        n = len(isConnected)
        parent = list(range(n))

        def find(i):
            if parent[i] != i:
                parent[i] = find(parent[i])
            return parent[i]

        def union(i, j):
            parent[find(i)] = find(j)

        for i in range(n):
            for j in range(i + 1, n):
                if isConnected[i][j] == 1:
                    union(i, j)
        return sum(parent[i] == i for i in range(n))
        